actions/dataset.go

```diff
@@ -12,56 +13,74 @@ import (
 	"github.com/qri-io/qri/base"
 	"github.com/qri-io/qri/p2p"
 	"github.com/qri-io/qri/repo"
+	"github.com/qri-io/qri/repo/profile"
 )
 
 // SaveDataset initializes a dataset from a dataset pointer and data file
 func SaveDataset(node *p2p.QriNode, dsp *dataset.DatasetPod, dryRun, pin bool) (ref repo.DatasetRef, body cafs.File, err error) {
 	var (
-		ds       *dataset.Dataset
-		bodyFile cafs.File
-		secrets  map[string]string
-		// NOTE - struct fields need to be instantiated to make assign set to
-		// new pointer values
-		userSet = &dataset.Dataset{
-			Commit:    &dataset.Commit{},
-			Meta:      &dataset.Meta{},
-			Structure: &dataset.Structure{},
-			Transform: &dataset.Transform{},
-			Viz:       &dataset.Viz{},
-		}
+		ds                       *dataset.Dataset
+		prevPath                 string
+		bodyFile, changeBodyFile cafs.File
+		secrets                  map[string]string
+		pro                      *profile.Profile
+		changes                  = &dataset.Dataset{}
+		r                        = node.Repo
 	)
 
+	// set ds to dataset head, or empty dataset if no history
+	ds, bodyFile, prevPath, err = base.PrepareDatasetSave(r, dsp.Peername, dsp.Name)
+	if err != nil {
+		return
+	}
+
 	if dryRun {
 		node.LocalStreams.Print("🏃🏽‍♀️ dry run\n")
-	}
 
-	// Determine if the save is creating a new dataset or updating an existing dataset by
-	// seeing if the name can canonicalize to a repo that we know about
-	lookup := &repo.DatasetRef{Name: dsp.Name, Peername: dsp.Peername}
-	err = repo.CanonicalizeDatasetRef(node.Repo, lookup)
-	if err == repo.ErrNotFound {
-		ds, bodyFile, secrets, err = base.PrepareDatasetNew(dsp)
+		pro, err = r.Profile()
 		if err != nil {
 			return
 		}
-	} else {
-		ds, bodyFile, secrets, err = base.PrepareDatasetSave(node.Repo, dsp)
+
+		// dry-runs store to an in-memory repo
+		r, err = repo.NewMemRepo(pro, cafs.NewMapstore(), profile.NewMemStore(), nil)
 		if err != nil {
 			return
 		}
 	}
 
-	mutateCheck := mutatedComponentsFunc(dsp)
-	userSet.Assign(ds)
+	if changeBodyFile, err = base.DatasetPodBodyFile(node.Repo.Store(), dsp); err == nil && changeBodyFile != nil {
+		dsp.BodyPath = ""
+		bodyFile = changeBodyFile
+	} else if err != nil {
+		return
+	}
+
+	if err = changes.Decode(dsp); err != nil {
+		return
+	}
+	ds.Assign(changes)
+	clearPaths(ds)
 
 	if ds.Transform != nil {
+		mutateCheck := mutatedComponentsFunc(dsp)
```

Comment:
Would be great to have a note here what exactly we are doing: making a record of all the parts that the datasetPod is changing, that way we can make sure the transform is not also changing them.

Part of the confusion, is that it isn't totally clear that `dsp` is a dataset made up of only the changes. `changes` makes sense, but maybe calling dsp `changesPod` might make this more clear?